carbon-design-system/carbon
# feat(icon-build-helpers): add module-info extension and next react builder

Closes #8179

This PR updates our `icon-build-helpers` with a new experimental builder for v11. This builder outputs a directory called `next` which contains a flat list of all of our icons. The intent is that teams would require icons from these sub-paths instead of importing from an entrypoint. This bundle is built using CommonJS, although this could change in the future.

<details>
<summary>Each icon has approximately the following shape before being compiled:</summary>

```jsx
import React from 'react';
import Icon from './Icon.js';

const ComponentName = React.forwardRef(
  function ComponentName({ children, size = 32, ... rest}, ref) {
    return (
      <Icon width={size} height={size} ref={ref} {...rest>
        <path d="..." />
        {children}
     </Icon>
  }
);

ComponentName.propTypes = {
  sizes: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default ComponentName;
```
</details>

<details>
<summary>For icons that have multiple asset sizes, we conditionally render each in the component based on an exact match:</summary>

```jsx
const ComponentName = React.forwardRef(
  function ComponentName({ children, size = 32, ... rest}, ref) {
    if (size === 16) {
      return (
        <Icon width={size} height={size} ref={ref} {...rest>
          <path d="..." />
          {children}
       </Icon>
      );
    }

    return (
      <Icon width={size} height={size} ref={ref} {...rest>
        <path d="..." />
        {children}
     </Icon>
  }
);
```
</details>

To accomplish this new format, we introduce a new extension in our metadata ecosystem: `module-info`. This extension iterates through our icons, extracting asset data, optimizing it, and converting it to an AST to be used by a builder. It also embeds information about the JavaScript module, like name and filepath.

#### Changelog

**New**

- Add new extension, `module-info`, to power the experimental builder for React icon components
- Add new builder, `reactNext`, for building our v11 set of icons

**Changed**

- Update the storybook example to work with Yarn 2 (sorry about the diff size 🤦 )

**Removed**

#### Testing / Reviewing

- Visit the storybook for icons-react
- Make sure the icon renders and no console errors are emitted
- Edit the index.stories.js file
  - Add a size prop, set it to 32 and verify the component renders
  - Change the size prop to 16 and verify the path data changes (menu uses multiple assets)
- Verify again that no prop warnings have been emitted
